controller_connections.go

```diff
@@ -90,19 +92,116 @@ func (c *controller) handleIncoming(con net.Conn) {
 		return
 	}
 
-	peer := newPeer(c.net, c.peerStatus, c.peerData)
-	// we are never expecting a reject-alternate for incoming connections
-	if _, err := peer.StartWithHandshake(ep, con, true); err != nil {
-		c.logger.WithError(err).Debugf("Handshake failed for address %s, stopping", ep)
-		peer.Stop()
+	// don't bother with alternatives for incoming connections
+	peer, _, err := c.handshake(host, con, true)
+
+	if err != nil {
+		c.logger.WithError(err).Debugf("inbound connection from %s failed handshake", host)
+		con.Close()
 		return
 	}
 
 	c.logger.Debugf("Incoming handshake success for peer %s, version %s", peer.Hash, peer.prot.Version())
+}
+
+func (c *controller) handshake(ip string, con net.Conn, incoming bool) (*Peer, []Endpoint, error) {
+	tmplogger := c.logger.WithField("addr", ip)
+	timeout := time.Now().Add(c.net.conf.HandshakeTimeout)
+
+	nonce := make([]byte, 8) // loopback detection
+	binary.LittleEndian.PutUint64(nonce, c.net.instanceID)
+
+	// upgrade connection to a metrics connection
+	metrics := NewMetricsReadWriter(con)
+
+	handshake := newHandshake(c.net.conf, nonce)
+	decoder := gob.NewDecoder(metrics) // pipe gob through the metrics writer
+	encoder := gob.NewEncoder(metrics)
+	con.SetWriteDeadline(timeout)
+	con.SetReadDeadline(timeout)
+
+	failfunc := func(err error) (*Peer, []Endpoint, error) {
+		tmplogger.WithError(err).Debug("Handshake failed")
+		con.Close()
+		return nil, nil, err
+	}
+
+	err := encoder.Encode(handshake)
+	if err != nil {
+		return failfunc(fmt.Errorf("Failed to send handshake to incoming connection"))
+	}
+
+	var reply Handshake
+	err = decoder.Decode(&reply)
+	if err != nil {
+		return failfunc(fmt.Errorf("Failed to read handshake from incoming connection"))
+	}
+
+	// check basic structure
+	if err = reply.Valid(c.net.conf); err != nil {
+		return failfunc(err)
+	}
+
+	endpoint, err := NewEndpoint(ip, reply.Header.PeerPort)
+	if err != nil {
+		return failfunc(fmt.Errorf("failed to create endpoint: %v", err))
+	}
+
+	if c.isBannedEndpoint(endpoint) {
+		return failfunc(fmt.Errorf("Peer %s is banned, disconnecting", endpoint))
+	}
 
-	if c.isBannedEndpoint(peer.Endpoint) {
-		c.logger.Debugf("Peer %s is banned, disconnecting", peer.Hash)
-		peer.Stop()
+	// loopback detection
+	if bytes.Equal(reply.Payload, nonce) {
+		return failfunc(fmt.Errorf("loopback"))
+	}
+
+	prot, err := c.determineProtocol(&reply, con, decoder, encoder)
+	if err != nil {
+		return failfunc(err)
+	}
+
+	if reply.Header.Type == TypeRejectAlternative {
```

Comment:
I'd add a short comment statement that this branch happens when we are in the outgoing case and that the node being dialled is full